### PR TITLE
Replace NodeJS v12 with NodeJS v16

### PR DESCRIPTION
**Pull request by KengoTODA**
Hi 👋 
Thanks for sharing your awesome GitHub Action! I'm using this to integrate SonarQube analysis with `pull_request_target` event.

I found that this action uses NodeJS v12 (Erbium) for now, [which will end its support next April](https://nodejs.org/en/about/releases/).

According to https://github.com/actions/runner/issues/772 now GitHub Actions can run on NodeJS v16 (Gallium). So this PR suggests migrating to NodeJS v16. It also contains several related changes:

* Apply the [recommended TSConfig settings for NodeJS v16](https://github.com/microsoft/TypeScript/wiki/Node-Target-Mapping#node-16)
* Bump up TypeScript to target "ES2021"
* Bump up ESLint family to support NodeJS 16 and latest TypeScript
* Change error handling because the [latest TypeScript treats caught object as `unknown` instead of `any`](https://devblogs.microsoft.com/typescript/announcing-typescript-4-4/#use-unknown-catch-variables)

Thanks for checking my PR! :wave:

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -16,13 +16,13 @@
...
-      "@typescript-eslint/ban-ts-ignore": "error",
+      "@typescript-eslint/ban-ts-comment": "error",
       "camelcase": "off",
-      "@typescript-eslint/camelcase": "error",
-      "@typescript-eslint/class-name-casing": "error",
+      "@typescript-eslint/consistent-type-assertions": "error",
+      "@typescript-eslint/consistent-type-definitions": "error",
       "@typescript-eslint/explicit-function-return-type": ["error", {"allowExpressions": true}],
       "@typescript-eslint/func-call-spacing": ["error", "never"],
-      "@typescript-eslint/generic-type-naming": ["error", "^[A-Z][A-Za-z]*$"],
```
Comment (KengoTODA) [inline .eslintrc.json:25]: `generic-type-naming` was replaced by `naming-convention`

https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0

File: .eslintrc.json [hunk 2]
```diff
@@ -16,13 +16,13 @@
       "@typescript-eslint/no-require-imports": "error",
       "@typescript-eslint/array-type": "error",
       "@typescript-eslint/await-thenable": "error",
-      "@typescript-eslint/ban-ts-ignore": "error",
+      "@typescript-eslint/ban-ts-comment": "error",
       "camelcase": "off",
-      "@typescript-eslint/camelcase": "error",
```
Comment (KengoTODA) [inline .eslintrc.json:21]: `camelcase` was replaced by `naming-convention`

https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0

File: .eslintrc.json [hunk 3]
```diff
@@ -16,13 +16,13 @@
       "@typescript-eslint/no-require-imports": "error",
       "@typescript-eslint/array-type": "error",
       "@typescript-eslint/await-thenable": "error",
-      "@typescript-eslint/ban-ts-ignore": "error",
```
Comment (KengoTODA) [inline .eslintrc.json:19]: `ban-ts-ignore` was replaced by `ban-ts-comment`

File: .eslintrc.json [hunk 4]
```diff
@@ -16,13 +16,13 @@
...
+      "@typescript-eslint/ban-ts-comment": "error",
       "camelcase": "off",
-      "@typescript-eslint/camelcase": "error",
-      "@typescript-eslint/class-name-casing": "error",
```
Comment (KengoTODA) [inline .eslintrc.json:22]: `class-name-casing` was replaced by `naming-convention`

https://github.com/typescript-eslint/typescript-eslint/releases/tag/v3.0.0

File: .eslintrc.json [hunk 5]
```diff
@@ -32,15 +32,13 @@
       "@typescript-eslint/no-misused-new": "error",
       "@typescript-eslint/no-namespace": "error",
       "@typescript-eslint/no-non-null-assertion": "warn",
-      "@typescript-eslint/no-object-literal-type-assertion": "error",
```
Comment (KengoTODA) [inline .eslintrc.json:35]: `no-object-literal-type-assertion` has merged into `consistent-type-assertions`.

https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0

File: .eslintrc.json [hunk 6]
```diff
@@ -32,15 +32,13 @@
...
       "@typescript-eslint/no-unnecessary-qualifier": "error",
       "@typescript-eslint/no-unnecessary-type-assertion": "error",
       "@typescript-eslint/no-useless-constructor": "error",
       "@typescript-eslint/no-var-requires": "error",
       "@typescript-eslint/prefer-for-of": "warn",
       "@typescript-eslint/prefer-function-type": "warn",
       "@typescript-eslint/prefer-includes": "error",
-      "@typescript-eslint/prefer-interface": "error",
```
Comment (KengoTODA) [inline .eslintrc.json:43]: Replaced by `consistent-type-definitions`.

https://github.com/typescript-eslint/typescript-eslint/releases/tag/v2.0.0

**Review by haya14busa — APPROVED** [review]
Thank you for your PR! 👍

**Pull request merged**